### PR TITLE
Fixed #25986 -- Fixed crash sending email with non-ASCII in local part of the address.

### PR DESCRIPTION
**Pull request by sergei-maertens**
On Python 3, sending e-mails failed for e-mail addresses containing
unicode characters due to the usage of the legacy python
email.utils.formataddr function. This has been fixed by using the
proper Address object if it's available (available with Python 3).

### CODE REVIEW

**Comment by sergei-maertens** [comment]
I know (at least) one test is failing for the SMTP backend on Python 3.

I belive the issue lies in `django.core.mail.EmailMessage.message`, where the headers are being set:

```
def message(self):
    encoding = self.encoding or settings.DEFAULT_CHARSET
    msg = SafeMIMEText(self.body, self.content_subtype, encoding)
    msg = self._create_message(msg)
    msg['Subject'] = self.subject
    msg['From'] = self.extra_headers.get('From', self.from_email)
```

If I'm not mistaken, these headers should also be sanitized with `django.core.mail.message.sanitize_address`, since in the SMTP backend they are sanitized.

Eventually, this bubbles up to the test case itself (tests.mail.tests.FakeSMTPServer.process_message) where a sanitized and non-sanitized version of the `from_email` are being compared. Due to the changes in this PR, there is an actual difference, even for ASCII local parts because the local part is now always MIME-word encoded.

TL;DR: shouldn't we also call `sanitize_address` in django.core.mail.EmailMessage.message?

**Comment by sergei-maertens** [comment]
I have added some remarks to the ticket as well - https://code.djangoproject.com/ticket/25986#comment:7

File: django/core/mail/message.py [hunk 1]
```diff
@@ -100,22 +100,62 @@ def forbid_multi_line_headers(name, val, encoding):
     return str(name), val
 
 
+def split_addr(addr, encoding):
+    """
+    Splits an address into local part and domain, properly encoded.
+    """
+    if '@' in addr:
+        localpart, domain = addr.split('@', 1)
+        localpart = Header(localpart, encoding).encode()
```
Comment (sergei-maertens) [inline django/core/mail/message.py]: This always encodes the local part with MIME word encoding, so `to@example.com` becomes something like `=?utf-8?q?to?=@example.com`. If we want to avoid that, we could do something like:
```
try:
    localpart.encode('ascii')
except UnicodeEncodeError:
    localpart = Header(localpart, encoding).encode()
```

this would make sure that simple localparts are kept simple and can be easily unit tested against by end users who test e-mail headers (`message.get('from')` etc.). That would also render a few unit test changes from Django unnecessary, making this PR a bit simpler. Any thoughts on this, anyone?

File: django/core/mail/message.py [hunk 2]
```diff
@@ -100,22 +100,62 @@ def forbid_multi_line_headers(name, val, encoding):
...
         addr.encode('ascii')
-    except UnicodeEncodeError:  # IDN
-        if '@' in addr:
-            localpart, domain = addr.split('@', 1)
-            localpart = str(Header(localpart, encoding))
-            domain = domain.encode('idna').decode('ascii')
+    except UnicodeEncodeError:  # IDN or non-ascii in the local part
+        localpart, domain = split_addr(addr, encoding)
+
+    # Do the feature test (Py2 vs. Py3)
```
Comment (timgraham) [inline django/core/mail/message.py]: Can we use `six.PY2` so it's easy to identify what code can be removed when dropping Python 2?

File: django/core/mail/message.py [hunk 3]
```diff
@@ -100,22 +100,62 @@ def forbid_multi_line_headers(name, val, encoding):
     return str(name), val
 
 
+def split_addr(addr, encoding):
+    """
+    Splits an address into local part and domain, properly encoded.
```
Comment (timgraham) [inline django/core/mail/message.py]: Use PEP257 verb style for new docstrings ("Split...."). Describe what "properly encoded" means?

File: django/core/mail/backends/smtp.py [hunk 4]
```diff
@@ -115,8 +115,9 @@ def _send(self, email_message):
         """A helper method that does the actual sending."""
         if not email_message.recipients():
             return False
-        from_email = sanitize_address(email_message.from_email, email_message.encoding)
-        recipients = [sanitize_address(addr, email_message.encoding)
+        encoding = email_message.encoding or settings.DEFAULT_CHARSET
+        from_email = sanitize_address(email_message.from_email, encoding)
+        recipients = [sanitize_address(addr, encoding)
                       for addr in email_message.recipients()]
```
Comment (timgraham) [inline django/core/mail/backends/smtp.py]: You can move this to the previous line (we are favoring slightly longer lines to avoid non-hanging indentation).

File: django/core/mail/message.py (hunk 1 above)
Comment (timgraham) [inline django/core/mail/message.py]: Sounds okay, let's include a comment about the reasoning.

File: django/core/mail/message.py [hunk 5]
```diff
@@ -100,22 +100,62 @@ def forbid_multi_line_headers(name, val, encoding):
...
+        localpart = Header(addr, encoding).encode()
+        domain = ''
+    return (localpart, domain)
+
+
 def sanitize_address(addr, encoding):
+    """
+    Formats a pair of (name, address) or an e-mail address string.
+
+    On Python 3, an `email.headerregistry.Address` object is used since
```
Comment (timgraham) [inline django/core/mail/message.py]: I would move these to inline comments near the imports as they can probably be mostly removed when dropping Python 2 and simplifying the code.

File: django/core/mail/message.py (hunk 3 above)
Comment (sergei-maertens) [inline django/core/mail/message.py]: Noted, will define properly encoded as 'MIME-word encoded` for the local part and `IDN-encoded` for the domain part. Do you want a reference to wiki/spec for what MIME-word encoded is in the docstring, or is the term enough?

File: django/core/mail/message.py (hunk 2 above)
Comment (sergei-maertens) [inline django/core/mail/message.py]: Probably, I'll check if it runs correctly on Py 3.3 but I think we're fine.

**Comment by sergei-maertens** [comment]
Tangent to this PR/ticket, I noticed yesterday that django's form field validation doesn't allow special characters. So the issues being fixed in this PR only happen if the e-mail data is entered through management commands or other data-import-like cases. That's probably why it hasn't shown up before.

However, since this PR fixes the bug, I feel like the e-mail validation regex needs adoption to allow accented letters as well (even though I'd rather not touch that thing with a long stick...). I'll make a ticket soon for this and refer to this ticket/PR.

**Comment by timgraham** [comment]
Actually, there's a ticket about simplifying the `EmailValidator` regex: https://code.djangoproject.com/ticket/26423. A comment there to add tests for that case when completing that work is probably sufficient.

File: django/core/mail/message.py (hunk 3 above)
Comment (timgraham) [inline django/core/mail/message.py]: I think googling that term will give a sufficient reference if needed.

File: django/core/mail/message.py (hunk 2 above)
Comment (timgraham) [inline django/core/mail/message.py]: master is only supporting Python 2.7 and 3.4+

**Comment by sergei-maertens** [comment]
@timgraham: remarks were fixed, up for review again

**Comment by timgraham** [comment]
merged in ec009ef1d8470f8fdb32802339fe2615e35887a1, thanks!
